ce/ce/registries/registries.ts

```diff
@@ -2,126 +2,238 @@
+  // multiple matches.
+  // we can't return a single artifact, we're going to have to throw.
+  fail(i`'${idOrShortName}' matched more than one result (${[...artifactRecords.map(each => each[0])].join(',')}).`);
+}
+
+export class RegistryDatabase {
+  #uriToRegistry: Map<string, Registry> = new Map();
+
+  getRegistryByUri(registryUri: string) {
+    return this.#uriToRegistry.get(registryUri);
   }
 
-  getRegistryName(registry: Registry): string {
-    for (const [name, reg] of this.registries) {
-      if (reg === registry && name.indexOf('://') === -1) {
-        return name;
-      }
+  entries() { return this.#uriToRegistry.values(); }
+
+  has(registryUri: string) { return this.#uriToRegistry.has(registryUri); }
+
+  add(uri: Uri, registry: Registry) {
+    const stringized = uri.toString();
+    if (this.#uriToRegistry.has(stringized)) {
+      throw new Error(`Duplicate registry add ${stringized}`);
     }
-    // ask the default registry?
-    return this.session.defaultRegistry.getRegistryName(registry);
+
+    this.#uriToRegistry.set(stringized, registry);
   }
 
-  getRegistry(id: string | Uri): Registry | undefined {
-    return this.registries.get(id.toString());
+  async loadRegistry(session: Session, locationUri: Uri): Promise<Registry> {
+    const locationUriStr = locationUri.toString();
+    const existingRegistry = this.#uriToRegistry.get(locationUriStr);
+    if (existingRegistry) {
+      return existingRegistry;
+    }
+
+    // not already loaded
+    let loaded: Registry;
+    switch (locationUri.scheme) {
+      case 'https':
+        loaded = new RemoteRegistry(session, locationUri);
+        break;
+
+      case 'file':
+        loaded = new LocalRegistry(session, locationUri);
+        break;
+
+      default:
+        throw new Error(i`Unsupported registry scheme '${locationUri.scheme}'`);
+    }
+
+    this.#uriToRegistry.set(locationUriStr, loaded);
+    await loaded.load();
+    return loaded;
   }
+}
 
-  has(registryName?: string) {
-    // only check for registries names not locations.
-    if (registryName && registryName.indexOf('://') === -1) {
-      for (const [name] of this.registries) {
-        if (name === registryName && name.indexOf('://') === -1) {
-          return true;
-        }
-      }
+// When a registry resolver is used to map a URI back to some form of for-display-purposes-only name.
+export interface RegistryDisplayContext {
+  getRegistryDisplayName(registry: Uri): string;
+}
+
+export class RegistryResolver implements RegistryDisplayContext {
+  readonly #database: RegistryDatabase;
+  readonly #knownUris: Set<string>;
+  readonly #uriToName: Map<string, string>;
+  readonly #nameToUri: Map<string, string>;
+
+  private addMapping(name: string, uri: string) {
+    this.#uriToName.set(uri, name);
+    this.#nameToUri.set(name, uri);
+  }
+
+  constructor(parent: RegistryDatabase | RegistryResolver) {
```

Comment:
It seems unfortunate that the constructor needs to be complicated just to support the "clone" operation in `combineWith`.

Reply:
idk, I think the clone operation makes sense to offer in general, and attempting to hoist it out either forces removing `readonly` from the set and maps, or giving up being able to call the ctor to clone those.